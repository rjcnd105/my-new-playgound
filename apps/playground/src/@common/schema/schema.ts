import { pipe } from '@effect/data/Function'
import * as Str from '@effect/data/String'
import * as S from '@effect/schema/Schema'

/**
 * 빈 문자열 스키마
 * @category primitives
 */
export const EmptyString = pipe(S.string, S.filter(Str.isEmpty))
