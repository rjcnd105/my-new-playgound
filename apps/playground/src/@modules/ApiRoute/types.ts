import type * as S from '@effect/schema/Schema'
import type { Schema } from '@effect/schema/Schema'
import type { Merge, Simplify } from 'type-fest'

import type { Method } from ':/@common/http/index.ts'
import type { MakePathParamsFromPath } from ':/@common/pathParams'

export interface InputApiRoute {
  path: string
  method: Method
  data?: S.Schema<any>
  params?: S.Schema<any>
  responseData?: S.Schema<any>
  headers?: HeadersInit
}

export interface InputApiRouteWithPathParams extends InputApiRoute {
  pathParams?: unknown
}
export type ApiRoute<T extends InputApiRouteWithPathParams = InputApiRouteWithPathParams> = Merge<
  T,
  MakePathParamsFromPath<T['path']>
>

// ApiRoute의 키 값들 추출, Schema일 경우 Schema To 타입 추출
export type ApiRouteAdapterOut<
  T extends ApiRoute<InputApiRouteWithPathParams>,
  PickKeysT extends keyof T = keyof T,
> = Simplify<{
  [K in PickKeysT as T[K] extends NonNullable<unknown> ? K : never]: T[K] extends S.Schema<
    any,
    infer Out
  >
    ? Out
    : T[K]
}>

// ApiRoute의 키 값들 추출, Schema일 경우 Schema To 타입 추출
export type ApiRouteAdapterInput<
  T extends InputApiRoute,
  PickKeysT extends keyof T = keyof T,
> = Simplify<{
  [K in PickKeysT as T[K] extends NonNullable<unknown> ? K : never]: T[K] extends S.Schema<
    infer In,
    any
  >
    ? In
    : T[K]
}>

// apiRoute에서의 ResponseData Output 타입 추출
// 서버로부터 받은 데이터가 parse 된 후의 값
export type ApiRouteOutput<T extends Pick<InputApiRoute, 'responseData'>> =
  T['responseData'] extends Schema<unknown> ? S.Schema.To<T['responseData']> : never

// 서버로 요청할 params 타입 추출
// parse 되기 전의 값
export type ApiRouteInputParams<T extends Pick<InputApiRoute, 'params'>> =
  T['params'] extends Schema<unknown> ? S.Schema.From<T['params']> : never

// 서버로 요청할 data 타입 추출
// parse 되기 전의 값
export type ApiRouteInputData<T extends Pick<InputApiRoute, 'data'>> =
  T['data'] extends Schema<unknown> ? S.Schema.From<T['data']> : never
