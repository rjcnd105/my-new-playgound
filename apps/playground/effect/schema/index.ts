import { pipe } from "@effect/data/Function";
import * as S from "@effect/schema/Schema";
import { transform } from "@effect/schema/Schema";
import z from "zod";

const zz = z.object({
  id: z.string().refine((a) => a === "dd"),
});

const a = pipe(
  S.struct({
    id: pipe(
      S.string,
      transform(
        S.number,
        (a) => a.length,
        (s) => `${s}`
      ),
      S.description("id")
    ),
  }),
  S.filter((a) => a.id > 2),
  S.message(() => "hihi")
);

const shape = S.getPropertySignatures(a);

const f = S.decode(shape.id)("aaa"); /*?*/

const b = pipe(S.string, S.brand("dd"));

S.encode(shape.id)(20); /*?*/

S.decode(a)({
  id: "ssss",
}); /*?*/

a.A; /*?*/
b.A; /*?*/
a.ast; /*?*/
const getProperty = <A, K extends keyof A>(
  s: S.Schema<A>,
  name: { [key in keyof A]: Schema<A[key]> }[K]
): {} => {
  if (s.ast._tag === "Refinement") {
  }
  if (s.ast._tag === "TypeLiteral") {
    return S.make(s.ast.propertySignatures[name]);
  }
  throw Error("Not a TypeLiteral");
};

getProperty(a); /*?*/
