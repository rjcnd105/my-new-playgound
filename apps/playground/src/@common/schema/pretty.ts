import type { ApiRoute } from ':/@modules/ApiRoute/types.ts'

export const ResponsePretty = <T extends ApiRoute>() => {}
