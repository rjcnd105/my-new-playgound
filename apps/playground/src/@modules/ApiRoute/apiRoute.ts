import type { ApiRoute, InputApiRoute } from ':/@modules/ApiRoute/types.js'

export const apiRoute = <const InputApiRouteT extends InputApiRoute>(route: InputApiRouteT) =>
  route as unknown as ApiRoute<InputApiRouteT>
