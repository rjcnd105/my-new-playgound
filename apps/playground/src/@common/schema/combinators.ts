import * as S from '@effect/schema/Schema'

import { EmptyString } from ':/@common/schema/schema.ts'

/**
 * 스키마에 빈 문자열을 허용하게 합니다.
 * @category combinators
 */
export const allowEmptyString = <I, A>(schema: S.Schema<I, A>) => S.union(EmptyString, schema)
