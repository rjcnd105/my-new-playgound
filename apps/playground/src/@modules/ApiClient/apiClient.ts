import { isNotNullable } from "@effect/data/Predicate";
import * as S from "@effect/schema/Schema";
import type { AxiosRequestConfig } from "axios";
import axios from "axios";

import { apiSuccessResponse } from ":/@common/http/response/ApiResponse.ts";
import { API_RESPONSE_SUCCESS } from ":/@common/http/response/types.ts";
import { parsePathParams } from ":/@common/pathParams/index.js";
import { throwError } from ":/@modules/ApiClient/middleware/throwError.ts";

import type {
  ApiRouteAdapterInput,
  InputApiRouteWithPathParams,
} from "../ApiRoute/types.ts";

export interface ApiClientOptions {
  isDecodeResponse?: boolean;
  isEncodeRequest?: boolean;
  responseDecoder?: <I, A>(data: I) => A;
  requestEncoder?: <T>(data: unknown) => T;
}

export interface Options extends AxiosRequestConfig, ApiClientOptions {}

export const makeApiClient = (options: Options) => {
  const axiosInstance = axios.create(options);

  return {
    axiosInstance,
    apiClient:
      <ApiRouteT extends InputApiRouteWithPathParams>(apiRoute: ApiRouteT) =>
      <
        InputT extends ApiRouteAdapterInput<
          ApiRouteT,
          "data" | "params" | "pathParams"
        >,
      >(
        input: InputT,
        opt: Omit<Options, "params" | "method"> | null = null,
      ) => {
        const mergeOpt: Options = Object.assign({}, apiRoute["options"], opt);

        // request 항목들을 encode할 것인지 여부
        const isEncodeRequest = !!mergeOpt.isEncodeRequest;
        // response 항목들을 decode할 것인지 여부
        const isDecodeResponse = !!mergeOpt.isDecodeResponse;

        const parsedPath = parsePathParams(apiRoute.path, input["pathParams"]);

        // requestParams
        const params =
          isNotNullable(apiRoute.params) && isEncodeRequest
            ? S.parseSync(apiRoute.params)(input["params"])
            : input["params"];

        // RequestBody(JSON)
        const data =
          isNotNullable(apiRoute.data) && isEncodeRequest
            ? S.parseSync(apiRoute.data)(input["data"])
            : input["data"];

        const finalOptions: Options = Object.assign(
          {
            method: apiRoute.method,
            params,
            data,
          },
          opt,
          {
            headers: Object.assign({}, apiRoute?.headers, opt?.["headers"]),
          },
        );

        const res = axiosInstance(parsedPath, finalOptions);

        return res
          .then((res) => {
            const parsedData = (apiRoute.responseData && isDecodeResponse
              ? S.parseSync(apiRoute.responseData)(res.data.data)
              : // @ts-ignore
                res.data.data) as unknown as S.Schema.To<
              ApiRouteT["responseData"]
            >;

            return apiSuccessResponse({
              type: API_RESPONSE_SUCCESS,
              code: res.data.code,
              data: parsedData,
              message: res.data.message,
              statusCode: res.status,
            });
          })
          .catch(throwError);
      },
  };
};
