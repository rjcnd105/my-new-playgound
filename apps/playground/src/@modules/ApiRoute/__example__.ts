import { pipe } from '@effect/data/Function'
import * as S from '@effect/schema/Schema'

import { apiRoute } from './apiRoute.ts'
import type {
  ApiRouteAdapterInput,
  ApiRouteAdapterOut,
  ApiRouteInputData,
  ApiRouteInputParams,
  ApiRouteOutput,
} from './types.ts'

const stringDateSchema = S.transform(
  S.string,
  S.Date,
  (v) => new Date(v),
  (v) => v.toString(),
)

const dateApi = apiRoute({
  method: 'POST',
  path: '/calendar/:year/:month/:day',
  params: S.struct({
    date: stringDateSchema,
  }),
  data: S.struct({
    code: S.string,
  }),
  responseData: S.struct({
    eventDates: pipe(
      S.array(S.string),
      S.transform(
        S.array(S.Date),
        (v) => v.map((d) => d.toString()),
        (v) => v.map((d) => new Date(d)),
      ),
    ),
  }),
})

export type DateApiRouteAdapterInput = ApiRouteAdapterInput<typeof dateApi>
export type DateApiRouteAdapterOut = ApiRouteAdapterOut<typeof dateApi>
export type DateApiOutput = ApiRouteOutput<typeof dateApi>
export type DateApiInputParams = ApiRouteInputParams<typeof dateApi>
export type DateApiInputData = ApiRouteInputData<typeof dateApi>
