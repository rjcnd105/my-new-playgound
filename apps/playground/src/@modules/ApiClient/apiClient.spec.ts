import * as S from '@effect/schema/Schema'
import { describe, it } from 'vitest'

import { makeApiClient } from ':/@modules/ApiClient/apiClient.ts'

import { apiRoute } from '../ApiRoute/apiRoute'

const { apiClient, axiosInstance } = makeApiClient({
  baseURL: 'https://api2-dev.mathflat.com',
  headers: {
    'x-platform': 'STUDENT',
  },
})

axiosInstance.defaults.timeout = 3000

describe('login api test', () => {
  const loginApiRoute = apiRoute({
    method: 'POST',
    path: 'login',
    data: S.struct({
      id: S.string,
      password: S.string,
    }),
    responseData: S.struct({
      id: S.string,
      userType: S.enums({ TEACHER: 'TEACHER', STUDENT: 'STUDENT' } as const),
      academyId: S.templateLiteral(S.literal('D'), S.string),
      relationId: S.templateLiteral(S.literal('T'), S.string),
      token: S.string,
    }),
    headers: {
      'x-my-header': 'test',
    },
  })

  it('스키마 성공', async ({ expect }) => {
    const login = apiClient(loginApiRoute)

    const loginResponseData = await login({
      data: {
        id: 'develop02',
        password: 'develop02',
      },
    })
    console.log('loginSuccess: ', loginResponseData)

    // expect(loginResponseData.kind).toBe(ResultTypes.API_RESPONSE_SUCCESS)
  })
  it('잘못 선언된 apiRoute로 인한 스키마 변환 실패', async ({ expect }) => {
    const wrongLoginApiRoute = apiRoute({
      method: 'POST',
      path: 'login',
      requestData: S.struct({
        id: S.string,
        password: S.string,
      }),
      responseData: S.struct({
        // 일부로 잘못 선언함
        idd: S.string,
        userType: S.enums({ TEACHER: 'TEACHER', STUDENT: 'STUDENT' }),
        academyId: S.templateLiteral(S.literal('D'), S.string),
        relationId: S.templateLiteral(S.literal('T'), S.string),
        token: S.string,
      }),
    })

    const wrongLoginResponseData = await apiClient(wrongLoginApiRoute)(
      {
        data: {
          id: 'develop02',
          password: 'develop02',
        },
      },
      { isDecodeResponse: true },
    ).catch(console.log)
  })
  it('서버 에러로 인한 실패', async ({ expect }) => {
    const failLogin = await apiClient(loginApiRoute)({
      data: {
        id: 'develop022222',
        password: 'develop02',
      },
    }).catch(console.log)
  })
})
