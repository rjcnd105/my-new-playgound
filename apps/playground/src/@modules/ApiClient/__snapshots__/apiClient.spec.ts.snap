// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`login api test > 스키마 Fail 테스트 1`] = `
{
  "code": null,
  "data": {
    "academyId": "D34345",
    "id": "develop02",
    "relationId": "T79880",
    "token": "480cb097-4d17-478f-be5f-9e008ee4543d",
    "userType": "TEACHER",
  },
  "message": null,
  "statusCode": 200,
  "type": "API_RESPONSE_SUCCESS",
}
`;
